crates/core/src/operations/cdc.rs

```diff
@@ -0,0 +1,454 @@
+            schema: self.schema(),
+            input: res,
+            sender: self.sender.clone(),
+        }))
+    }
+
+    fn statistics(&self) -> DataFusionResult<datafusion_common::Statistics> {
+        self.parent.statistics()
+    }
+
+    fn with_new_children(
+        self: Arc<Self>,
+        children: Vec<Arc<dyn ExecutionPlan>>,
+    ) -> datafusion_common::Result<Arc<dyn ExecutionPlan>> {
+        if let Some(parent) = children.first() {
+            Ok(Arc::new(CDCObserver {
+                id: self.id.clone(),
+                sender: self.sender.clone(),
+                parent: parent.clone(),
+            }))
+        } else {
+            Err(datafusion_common::DataFusionError::Internal(
+                "Failed to handle CDCObserver".into(),
+            ))
+        }
+    }
+
+    fn metrics(&self) -> Option<MetricsSet> {
+        self.parent.metrics()
+    }
+}
+
+/// The CDCObserverStream simply acts to help observe the stream of data being
+/// read by DataFusion to capture the pre-image versions of data
+pub(crate) struct CDCObserverStream {
+    schema: SchemaRef,
+    input: SendableRecordBatchStream,
+    sender: Sender<RecordBatch>,
+}
+
+impl Stream for CDCObserverStream {
+    type Item = DataFusionResult<RecordBatch>;
+
+    fn poll_next(
+        mut self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Option<Self::Item>> {
+        self.input.poll_next_unpin(cx).map(|x| match x {
+            Some(Ok(batch)) => {
+                let _ = self.sender.try_send(batch.clone());
+                Some(Ok(batch))
+            }
+            other => other,
+        })
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        self.input.size_hint()
+    }
+}
+
+impl RecordBatchStream for CDCObserverStream {
+    fn schema(&self) -> SchemaRef {
+        self.schema.clone()
+    }
+}
+
+///
+/// Return true if the specified table is capable of writing Change Data files
+///
+/// From the Protocol:
+///
+/// > For Writer Versions 4 up to 6, all writers must respect the delta.enableChangeDataFeed
+/// > configuration flag in the metadata of the table. When delta.enableChangeDataFeed is true,
+/// > writers must produce the relevant AddCDCFile's for any operation that changes data, as
+/// > specified in Change Data Files.
+/// >
+/// > For Writer Version 7, all writers must respect the delta.enableChangeDataFeed configuration flag in
+/// > the metadata of the table only if the feature changeDataFeed exists in the table protocol's
+/// > writerFeatures.
+pub(crate) fn should_write_cdc(snapshot: &DeltaTableState) -> DeltaResult<bool> {
+    if let Some(features) = &snapshot.protocol().writer_features {
+        // Features should only exist at writer version 7 but to avoid cases where
+        // the Option<HashSet<T>> can get filled with an empty set, checking for the value
+        // explicitly
+        if snapshot.protocol().min_writer_version == 7
```

Comment:
There should be an else_if that returns true if it's V7 and ChangeDateFeed.

It now just goes into the configuration grab at the end, but it's possible that the WriterFeature is enabled and someone tried to disable it through the config, which doesn't have an effect

Reply:
I believe that the `table_config().enable_change_data_feed()` function's value below would be correct for that case wouldn't it?

Reply:
Not entirely, it's possible to have delta.enableChangeDataFeed = False in the config, even though the WriterFeature ChangeDateFeed is set in the protocol features of the table.

That's why we need to do:

If writer >= 7 {
   If writer_features.contains(cdf) {
        True
    } Else {
  False 
}}

Because >=7 we should only look at features

Reply:
Hmmmmm

Protocol states it does still need to look at the config: 

"For Writer Version 7, all writers must respect the delta.enableChangeDataFeed configuration flag in the metadata of the table only if the feature changeDataFeed exists in the table protocol's writerFeatures." 

I guess I misunderstood what spark was doing then :) forget everything I said ^^